lib/logger_splunk_backend.ex

```diff
@@ -5,140 +5,304 @@ defmodule Logger.Backend.Splunk do
   end
 
   defp format_event(level, msg, ts, md, %{compiled_format: format, metadata: keys}) do
     Logger.Formatter.format(format, level, msg, ts, take_metadata(md, keys))
   end
 
-  def maybe_send(%{buffer_size: bs, max_buffer: mb} = state) when bs >= mb do
-    state.connector.transmit(state.buffer, state.host, state.token)
-    %{state |
-      buffer: [],
-      buffer_size: 0
-    }
+  defp maybe_send(%{response: nil, buffer_size: bs} = state) when bs != 0 do
+    case send_to_splunk(state) do
+      {:ok, response} ->
+        %{state | response: response, output: state.buffer, buffer: [], buffer_size: 0}
+
+      {:error, %{reason: reason}} ->
+        IO.puts(state.error_device, [
+          "ERROR unable to connect to Splunk: ",
+          inspect(reason),
+          "\n",
+          "ERROR sent: ",
+          state.buffer
+        ])
+
+        %{state | buffer: [], buffer_size: 0}
+    end
   end
-  def maybe_send(state) do
+
+  defp maybe_send(state) do
     state
   end
 
+  defp send_to_splunk(state) do
+    headers = [{"Authorization", "Splunk #{state.token}"}, {"Content-Type", "application/json"}]
+    opts = [hackney: [pool: :logger_splunk_backend], stream_to: self()]
+    HTTPoison.post(state.host, IO.iodata_to_binary(state.buffer), headers, opts)
+  end
+
+  defp await_response(%{response: %{id: ref}} = state) do
+    receive do
+      %{id: ^ref} = response ->
+        state = handle_http_response(response, state)
+        await_response(state)
+    end
+  end
+
+  defp await_response(state) do
+    maybe_send(state)
+  end
+
+  defp take_metadata(metadata, :all) do
+    Keyword.delete(metadata, :crash_reason)
+  end
+
   defp take_metadata(metadata, keys) do
     Enum.reduce(keys, [], fn key, acc ->
       case Keyword.fetch(metadata, key) do
         {:ok, val} -> [{key, val} | acc]
-        :error     -> acc
+        :error -> acc
       end
-    end) |> Enum.reverse()
+    end)
   end
 
   defp configure(opts, state) do
     name = state.name
     env = Application.get_env(:logger, name, [])
     opts = Keyword.merge(env, opts)
     Application.put_env(:logger, name, opts)
-    connector = Keyword.get(opts, :connector, state.connector)
     host = Keyword.get(opts, :host, state.host)
     level = Keyword.get(opts, :level, state.level)
-    metadata = Keyword.get(opts, :metadata, state.metadata)
     format = Keyword.get(opts, :format, state.format)
     max_buffer = Keyword.get(opts, :max_buffer, state.max_buffer)
+    error_device = Keyword.get(opts, :error_device, state.error_device)
 
-    %{state |
-      connector: connector,
-      host: host,
-      level: level,
-      format: format,
-      max_buffer: max_buffer,
-      compiled_format: Logger.Formatter.compile(format),
-      metadata: metadata,
-      token: token(Keyword.get(opts, :token, ""))
+    metadata =
+      case Keyword.fetch(opts, :metadata) do
+        {:ok, :all} -> :all
+        {:ok, metadata} when is_list(metadata) -> Enum.reverse(metadata)
```

Comment:
Why does it need to be reversed?

Reply:
It's reversed here so that the actual metadata processing in https://github.com/mbta/logger_splunk_backend/pull/6/files#diff-422fe5b31c1b6f41cc85cb125846e02cR248-R255 doesn't need to do a reverse on each call.